File: social-outing-planner/server/routes/content.py
from flask import Blueprint, request, jsonify
from server.config import Config
from server.storage import JsonStorage
from server.models import POI, AOI
from server.utils import generate_id, token_required

content_bp = Blueprint('content', __name__)

# POI Routes
@content_bp.route('/pois', methods=['GET'])
@token_required
def get_pois(current_user):
    try:
        pois = JsonStorage.read(Config.POIS_FILE)
        user_pois = [poi for poi in pois if poi.get('userId') == current_user['userId']]
        return jsonify(user_pois)
    except Exception as error:
        return jsonify({'error': str(error)}), 500

@content_bp.route('/pois', methods=['POST'])
@token_required
def create_poi(current_user):
    try:
        data = request.get_json()
        pois = JsonStorage.read(Config.POIS_FILE)
        
        poi_data = {
            '_id': generate_id(),
            **data,
            'userId': current_user['userId']
        }
        poi = POI.from_dict(poi_data)
        
        pois.append(poi.to_dict())
        JsonStorage.write(Config.POIS_FILE, pois)
        
        return jsonify(poi.to_dict())
    except Exception as error:
        return jsonify({'error': str(error)}), 400

@content_bp.route('/pois/<poi_id>', methods=['DELETE'])
@token_required
def delete_poi(current_user, poi_id):
    try:
        pois = JsonStorage.read(Config.POIS_FILE)
        filtered_pois = [poi for poi in pois if not (poi.get('_id') == poi_id and poi.get('userId') == current_user['userId'])]
        JsonStorage.write(Config.POIS_FILE, filtered_pois)
        
        return jsonify({'message': 'POI deleted successfully'})
    except Exception as error:
        return jsonify({'error': str(error)}), 500

# AOI Routes
@content_bp.route('/aois', methods=['GET'])
@token_required
def get_aois(current_user):
    try:
        aois = JsonStorage.read(Config.AOIS_FILE)
        user_aois = [aoi for aoi in aois if aoi.get('userId') == current_user['userId']]
        return jsonify(user_aois)
    except Exception as error:
        return jsonify({'error': str(error)}), 500

@content_bp.route('/aois', methods=['POST'])
@token_required
def create_aoi(current_user):
    try:
        data = request.get_json()
        aois = JsonStorage.read(Config.AOIS_FILE)
        
        aoi_data = {
            '_id': generate_id(),
            **data,
            'userId': current_user['userId']
        }
        aoi = AOI.from_dict(aoi_data)
        
        aois.append(aoi.to_dict())
        JsonStorage.write(Config.AOIS_FILE, aois)
        
        return jsonify(aoi.to_dict())
    except Exception as error:
        return jsonify({'error': str(error)}), 400

@content_bp.route('/aois/<aoi_id>', methods=['DELETE'])
@token_required
def delete_aoi(current_user, aoi_id):
    try:
        aois = JsonStorage.read(Config.AOIS_FILE)
        filtered_aois = [aoi for aoi in aois if not (aoi.get('_id') == aoi_id and aoi.get('userId') == current_user['userId'])]
        JsonStorage.write(Config.AOIS_FILE, filtered_aois)
        
        return jsonify({'message': 'AOI deleted successfully'})
    except Exception as error:
        return jsonify({'error': str(error)}), 500

File: social-outing-planner/rust-server/src/routes/content.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::Deserialize;
use crate::{
    models::{AOI, POI},
    state::AppState,
    utils::{generate_id, AuthUser},
};

// DTOs
#[derive(Deserialize)]
pub struct CreatePOIRequest {
    #[serde(default)]
    name: String,
    #[serde(default)]
    location: String,
    #[serde(default)]
    description: String,
    #[serde(default)]
    category: String,
}

#[derive(Deserialize)]
pub struct CreateAOIRequest {
    #[serde(default)]
    name: String,
    #[serde(default)]
    description: String,
    #[serde(default)]
    duration: Option<String>,
    #[serde(default)]
    category: String,
}

// POI Handlers
pub async fn get_pois(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let pois: Vec<POI> = state.pois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    let user_pois: Vec<POI> = pois.into_iter().filter(|p| p.user_id == auth.user_id).collect();
    Ok(Json(user_pois))
}

pub async fn create_poi(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(payload): Json<CreatePOIRequest>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let mut pois: Vec<POI> = state.pois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let new_poi = POI {
        id: generate_id(),
        user_id: auth.user_id,
        name: payload.name,
        location: payload.location,
        description: payload.description,
        category: payload.category,
        created_at: chrono::Utc::now().to_rfc3339(),
    };

    pois.push(new_poi.clone());
    state.pois_store.write(&pois).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(new_poi))
}

pub async fn delete_poi(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(poi_id): Path<String>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let pois: Vec<POI> = state.pois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    
    let filtered_pois: Vec<POI> = pois.into_iter()
        .filter(|p| !(p.id == poi_id && p.user_id == auth.user_id))
        .collect();
    
    // Note: This logic follows Python: if POI doesn't exist or not owned, it effectively does nothing but writes back.
    // Ideally we should return 404/403 but Python impl just filters.
    
    state.pois_store.write(&filtered_pois).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "message": "POI deleted successfully" })))
}

// AOI Handlers
pub async fn get_aois(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let aois: Vec<AOI> = state.aois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    let user_aois: Vec<AOI> = aois.into_iter().filter(|a| a.user_id == auth.user_id).collect();
    Ok(Json(user_aois))
}

pub async fn create_aoi(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(payload): Json<CreateAOIRequest>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let mut aois: Vec<AOI> = state.aois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let new_aoi = AOI {
        id: generate_id(),
        user_id: auth.user_id,
        name: payload.name,
        description: payload.description,
        duration: payload.duration.unwrap_or_default(),
        category: payload.category,
        created_at: chrono::Utc::now().to_rfc3339(),
    };

    aois.push(new_aoi.clone());
    state.aois_store.write(&aois).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(new_aoi))
}

pub async fn delete_aoi(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(aoi_id): Path<String>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let aois: Vec<AOI> = state.aois_store.read().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    
    let filtered_aois: Vec<AOI> = aois.into_iter()
        .filter(|a| !(a.id == aoi_id && a.user_id == auth.user_id))
        .collect();
    
    state.aois_store.write(&filtered_aois).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "message": "AOI deleted successfully" })))
}
